### PR TITLE
Implement RFC style for match pattern

### LINKED ISSUE #1807: Match expansion when arms have many variants
I had this match block:

```rust
match &*ext.to_lowercase() {
    "bmp" | "gif" | "ico" | "jpe" | "jpeg" | "jpg" | "png" | "svg" | "tif" | "tiff" |
     "webp" | "xcf" | "psd" | "ai" => FileType::Image,

    "3g2" | "3gp" | "avi" | "divx" | "flv" | "mov" | "mp4" | "mp4v" | "mpa" | "mpe" |
    "mpeg" | "ogv" | "qt" | "webm" | "wmv" => FileType::Video,

    "7z" | "rar" | "tgz" | "gz" | "zip" | "tar" | "xz" | "dmg" | "iso" | "lzma" |
    "tlz" | "bz2" | "tbz2" | "z" | "deb" | "rpm" | "jar" => FileType::Archive,

    "aac" | "flac" | "ogg" | "au" | "mid" | "midi" | "mp3" | "mpc" | "ra" | "wav" |
    "axa" | "oga" | "spz" | "xspf" | "wma" | "m4a" => FileType::Audio,

    _ => FileType::Unknown,
}
```

and it was reformatted, one-match per arm, into [this 64-line gist](https://gist.github.com/anowell/27388ed235e5ad529da0a634ac40b771) which seems less than ideal. In general, I like rustfmt's current match formatting, but this feels like a case where insert, reflow, and deal with the strange diff seems better than the really long expanded form.

### PR DESCRIPTION
**Pull request by topecongiro**
This PR implements RFC style for match pattern:

1. Break before `|` when using multiple lines (https://github.com/rust-lang-nursery/rustfmt/pull/2161/commits/09d54512da002e9695402b88fda63bcc49f6b762 ~ https://github.com/rust-lang-nursery/rustfmt/pull/2161/commits/1c702aa252dc048dd26bf0eea828aafbd52795d5)
2. Use mixed layout when all patterns are 'short' (https://github.com/rust-lang-nursery/rustfmt/pull/2161/commits/3bd1cc4fe072ccc6bd53ee57c90af3debca377fc ~ https://github.com/rust-lang-nursery/rustfmt/pull/2161/commits/95d7619e345ac6fa997297af9565bfd44ef9e855)

Closes #1807.

### CODE REVIEW

File: src/expr.rs [hunk 1]
```diff
@@ -1411,6 +1411,38 @@ fn rewrite_match_arm(
...
+///     - single token
+///     - `&[single-line, ntp]`
+///
+/// [small]:
+///     - `[small, ntp]`
+///     - unary tuple constructor `([small, ntp])`
+///     - `&[small]`
+fn is_short_pattern(pat: &ast::Pat, pat_str: &str) -> bool {
+    // We also require that the pattern is reasonably 'small' with its literal width.
+    pat_str.len() <= 20 && !pat_str.contains("\n") && is_short_pattern_inner(pat)
```
Comment (nrc) [inline src/expr.rs]: Do you think it is necessary to have the length check?

Comment (topecongiro) [reply src/expr.rs]: I added the length check to work around some extreme cases:

```diff
Mismatch at tests/target/configs-match_pattern_separator_break_point-Front.rs:3:
 ⏎
 fn main() {⏎
     match lorem {⏎
-        aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa⏎
-        | bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb⏎                                                                              
-        | cccccccccccccccccccccccccccccccccccccccc⏎                                                                              
-        | dddddddddddddddddddddddddddddddddddddddd => (),⏎                                                                       
+        aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa | bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb⏎
+        | cccccccccccccccccccccccccccccccccccccccc | dddddddddddddddddddddddddddddddddddddddd => (),⏎                            
         _ => (),⏎
     }⏎
 }⏎

Mismatch at tests/source/configs-match_pattern_separator_break_point-Front.rs:3:
 ⏎
 fn main() {⏎
     match lorem {⏎
-        aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa⏎
-        | bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb⏎                                                                              
-        | cccccccccccccccccccccccccccccccccccccccc⏎                                                                              
-        | dddddddddddddddddddddddddddddddddddddddd => (),⏎                                                                       
+        aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa | bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb⏎
+        | cccccccccccccccccccccccccccccccccccccccc | dddddddddddddddddddddddddddddddddddddddd => (),⏎                            
         _ => (),⏎
     }⏎
 }⏎

Mismatch at tests/source/match.rs:19:
         Pattern1 | Pattern2 | Pattern3 => false,⏎
         Paternnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnn⏎
         | Paternnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnn => blah,⏎
-        Patternnnnnnnnnnnnnnnnnnn⏎
-        | Patternnnnnnnnnnnnnnnnnnn⏎                                                                                             
-        | Patternnnnnnnnnnnnnnnnnnn⏎                                                                                             
+        Patternnnnnnnnnnnnnnnnnnn | Patternnnnnnnnnnnnnnnnnnn | Patternnnnnnnnnnnnnnnnnnn⏎
         | Patternnnnnnnnnnnnnnnnnnn => meh,⏎
 ⏎
         Patternnnnnnnnnnnnnnnnnnn | Patternnnnnnnnnnnnnnnnnnn if looooooooooooooooooong_guard => {⏎

Mismatch at tests/target/match.rs:19:
         Pattern1 | Pattern2 | Pattern3 => false,⏎
         Paternnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnn⏎
         | Paternnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnn => blah,⏎
-        Patternnnnnnnnnnnnnnnnnnn⏎
-        | Patternnnnnnnnnnnnnnnnnnn⏎                                                                                             
-        | Patternnnnnnnnnnnnnnnnnnn⏎                                                                                             
+        Patternnnnnnnnnnnnnnnnnnn | Patternnnnnnnnnnnnnnnnnnn | Patternnnnnnnnnnnnnnnnnnn⏎
         | Patternnnnnnnnnnnnnnnnnnn => meh,⏎
 ⏎
         Patternnnnnnnnnnnnnnnnnnn | Patternnnnnnnnnnnnnnnnnnn if looooooooooooooooooong_guard => {⏎
test self_tests ... ok

Mismatch at tests/source/expr.rs:402:
     ];⏎
 ⏎
     match x {⏎
-        xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx⏎
-        | yyyyyyyyyyyyyyyyyyyyyyyyyyyyyy⏎                                                                                        
+        xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx | yyyyyyyyyyyyyyyyyyyyyyyyyyyyyy⏎
         | zzzzzzzzzzzzzzzzzzzzzzzzzzzzzz => foo(a, b, c),⏎
         _ => bar(),⏎
     };⏎

Mismatch at tests/target/expr.rs:402:
     ];⏎
 ⏎
     match x {⏎
-        xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx⏎
-        | yyyyyyyyyyyyyyyyyyyyyyyyyyyyyy⏎
+        xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx | yyyyyyyyyyyyyyyyyyyyyyyyyyyyyy⏎
         | zzzzzzzzzzzzzzzzzzzzzzzzzzzzzz => foo(a, b, c),⏎
         _ => bar(),⏎
     };⏎
```

**Comment by topecongiro** [comment]
I don't think the absolute length check is a good idea, but added it to work around the extreme cases like above. I prefer to remove the length check if possible, it's just that someone using long constants may get mildly punished.

File: src/expr.rs (hunk 1 above)
Comment (nrc) [reply src/expr.rs]: All these examples look like they should be caught by the `!pat_str.contains("\n")` clause, am I misunderstanding how that works?

Comment (topecongiro) [reply src/expr.rs]: Sorry for ambiguity. The examples show the diffs when we remove `pat_str.len() <= 20` from the PR. 

Without the length check, we use mixed layout (`+` in diffs). With the length check, we use vertical layout (`-` in diffs). 

My first idea was to use vertical layout for excessively long patterns.

Comment (nrc) [reply src/expr.rs]: AH, I see, thanks for explaining!

**Pull request merged**